commitment_scheme/src/packer_hasher.rs

```diff
@@ -0,0 +1,297 @@
+                }
+            }
+
+            // Store the results in the returned map.
+            let hash_array = self.pack_and_hash_internal(&packed_elements, is_merkle_layer);
+            hashed_packages.insert(package, hash_array);
+        }
+
+        hashed_packages
+    }
+}
+
+/// Computes the number elements in a package.
+///
+/// # Arguments
+///
+/// - `size_of_element`: length of element in bytes.
+/// - `size_of_package`: length of package in bytes.
+/// - `max_n_elements`: number of elements to be grouped in a package.
+///
+/// # Returns
+///
+/// Returns the number elements in a package.
+fn compute_num_elements_in_package(
+    size_of_element: usize,
+    size_of_package: usize,
+    max_n_elements: usize,
+) -> usize {
+    assert!(
+        size_of_element > 0,
+        "An element must be at least of length 1 byte."
+    );
+
+    if size_of_element >= size_of_package {
+        return 1;
+    }
+
+    let elements_fit_in_package = (size_of_package - 1) / size_of_element + 1;
+    let log2_elements_fit_in_package = (elements_fit_in_package as f64).log2().ceil() as u32;
+    usize::min(1 << log2_elements_fit_in_package, max_n_elements)
+}
+
+/// Given a sequence of bytes, partitions the sequence to n_elements equal sub-sequences,
+/// hashing each separately, and returning the resulting sequence of hashes as vector of bytes.
+///
+/// # Arguments
+///
+/// - `data`: bytes to be hashed.
+/// - `n_elements`: number of elements.
+///
+/// # Returns
+///
+/// Returns the resulting sequence of hashes as vector of bytes.
+pub fn hash_elements<F: PrimeField, H: Hasher<F, Output = Vec<u8>>>(
+    data: &[u8],
+    n_elements: usize,
+) -> Vec<u8> {
+    // If n_elements == 0 and data is empty, return an empty vector.
+    if n_elements == 0 && data.is_empty() {
+        return Vec::new();
+    }
+
+    let element_size = data.len() / n_elements;
+    let mut res = Vec::with_capacity(n_elements * H::DIGEST_NUM_BYTES);
+    let mut pos = 0;
+
+    for _ in 0..n_elements {
+        let end_pos = pos + element_size;
+        let chunk = &data[pos..end_pos];
+        let mut hash_as_bytes_array = H::hash_bytes(chunk).to_vec();
+        res.append(&mut hash_as_bytes_array);
+        pos = end_pos;
+    }
+
+    res
+}
+
+/// Given a sequence of bytes, partitions the sequence to elements and hashes each pair separately.
+///
+/// # Arguments
+///
+/// - `data`: bytes to be hashed.
+///
+/// # Returns
+///
+/// Returns the resulting sequence of hashes as vector of bytes.
+pub fn hash_elements_two_to_one<F: PrimeField, H: Hasher<F, Output = Vec<u8>>>(
+    data: &[u8],
+) -> Vec<u8> {
+    // If data is empty, return an empty vector.
+    if data.is_empty() {
+        return Vec::new();
+    }
+
+    // Each element in the next layer is a hash of 2 elements in the current layer.
+    let elements_to_hash_size = 2 * H::DIGEST_NUM_BYTES;
+    let n_elements_next_layer = data.len() / elements_to_hash_size;
```

Comment:
The following assert before calling this functions ensures that it divides.

https://github.com/varunthakore/pumice/blob/7fbe77735322bd9235b449969fb03a1399b797dd/commitment_scheme/src/packer_hasher.rs#L66-L69

Reply:
Maybe I'm missing context, but if, e.g. 

- data.len() = 4 * H::DIGEST_NUM_BYTES + 1
- n_elements_in_package = 2
- size_of_element = H::DIGEST_NUM_BYTES

Then:

- n_elements_in_data = data.len() / size_of_element = 4
- n_packages = n_elements_in_data / n_elements_in_package = 2
- data.len() / n_packages = 2 * H::DIGEST_NUM_BYTES

So the check passes because the remainder data.len() % n_packages is discarded.

Due to rounding it should probably be:

```
assert_eq!(data.len(), n_packages * 2 * H::DIGEST_NUM_BYTES, "Data size is wrong.");
```

In general when checking conditions like this try to avoid division: division is not defined over the ring of integers so you run into issues like this.